src/components/tl/ucp/allreduce/allreduce_sliding_window.c

```diff
@@ -11,80 +11,631 @@
+
+            if (put_window_size <= 0)
+                put_window_size = host_team_size;
+
+            ucc_assert(put_window_size > 0);
+
+            window = ucc_min(put_window_size,
+                             host_team_size - pipe->posted_put);
+
+            for (i = 0; i < window; i++) {
+                dst_rank = pipe->dst_rank;
+                src_addr = accbuf->buf;
+                dst_addr = PTR_OFFSET(
+                            task->allreduce_sliding_window.bufs->rbufs[dst_rank],
+                            put_offset);
+                put_idx = pipe->posted_put %
+                          put_window_size;
+
+                if (task->allreduce_sliding_window.put_requests[put_idx] !=
+                    NULL) {
+                    // We've already posted a put at this index that didn't yet
+                    // complete, left this function and came back. Skip to check
+                    // whether this request finished instead of overwriting it
+                    // with another put
+                    break;
+                }
+
+                ucp_worker_fence(tl_ctx->worker.ucp_worker);
+                ucc_tl_ucp_get_ep(tl_team, dst_rank, &ep);
+                task->allreduce_sliding_window.put_requests[put_idx] = 
+                    ucp_put_nbx(
+                        ep, src_addr,
+                        data_size, (uint64_t)dst_addr,
+                        task->allreduce_sliding_window.bufs->dst_rkeys[dst_rank],
+                        &req_param);
+
+                pipe->posted_put++;
+                pipe->dst_rank = (dst_rank + 1) % host_team_size;
+            }
+
+            for (i = pipe->done_put; i < pipe->posted_put; i++) {
+                put_idx = i % put_window_size;
+                request = task->allreduce_sliding_window.put_requests[put_idx];
+
+                // These are fenced, so if the first fails, the proceding will
+                // too
+                if (ucc_tl_ucp_allreduce_sliding_window_req_test(
+                        request, task) != UCC_OK)
+                    break;
+
+                if (request != NULL)
+                    ucp_request_free(request);
+                task->allreduce_sliding_window.put_requests[put_idx] = NULL;
+                pipe->done_put++;
+            }
+
+            if (pipe->done_put == host_team_size) {
+                ucc_assert(pipe->avail_buffs == pipe->num_buffers);
+                ucc_assert(pipe->done_get == host_team_size);
+                ucc_assert(pipe->done_red == host_team_size - 1);
+                ucc_assert(pipe->done_put == host_team_size);
+
+                pipe->count_serviced += accbuf->count;
+
+                ucc_tl_ucp_allreduce_sliding_window_reset_buf(accbuf);
+                pipe->done_get = 0;
+                pipe->done_red = pipe->done_put = pipe->posted_put = 0;
+
+                memset(task->allreduce_sliding_window.put_requests, 0,
```

Comment:
Actually here it is required since at the start of the algorithm they need to be initialized to NULL. The other memset in the algorithm was not required. I removed it.